### PR TITLE
Camera: Add movement items to view menu

### LINKED ISSUE #70: Improve camera control
The current camera control uses vanilla `Qt3DExtras::QOrbitCameraController` and should be improved.  The camera control is a critical component in the viewer.  It at least needs the following functions and more:

* Resetting.
* WSAD.
* Change of origin.

----

Amended on 27th July 2024

- [x] "Up" and "w" should move toward the top of the window.  "Down" and "s" should move toward the bottom of the window.  "Q" and "e" should change to move along the z axis (perpendicular to the window).
  - I am not sure about the binding to page up/down.  We can use discussions.
  - Implemented with PR #408.
- [x] Camera reset/original position to be configurable and exposed through Python API.  More information: https://github.com/solvcon/modmesh/pull/408#discussion_r1718535001
  - Done with PR #418

### PR DESCRIPTION
**Pull request by Tucchhaa**
https://github.com/solvcon/modmesh/issues/70

In this PR:
* Now Ctrl+W/Arrow up is to move camera forward and Ctrl+S/Arrow down to backward. Q/E and Page Up/Down keys are removed
* Added items to View submenu to control camera translation and rotation
* Refactored resetCamera function
* Fixed the error when there are no subwindows open, but resetCamera is pressed

Screenshot 2024-08-14 at 20 54 01

### CODE REVIEW

File: cpp/modmesh/view/RManager.cpp
Comment (tigercosmos) [inline cpp/modmesh/view/RManager.cpp]: You don't need to check `std::bad_cast` here; instead, you should check if it's `null_ptr`.

According to the [doc](https://en.cppreference.com/w/cpp/language/dynamic_cast):
> If target-type is a pointer type, the result is the null pointer value of target-type.
If target-type is a reference type, an exception of a type that would match a [handler](https://en.cppreference.com/w/cpp/language/catch) of type [std::bad_cast](https://en.cppreference.com/w/cpp/types/bad_cast) is thrown.

**Comment by yungyuc** [comment]
@Tucchhaa Is this PR still a draft?  If so, feel free to convert it to a draft.

If it is supposed to be reviewed, please add review comments at the code to highlight the points to be reviewed, and a clear summary in the description.  Past PRs have examples, e.g., PR #356 and #398.  Also, please make CI clean before requesting for review.

**Comment by Tucchhaa** [comment]
@yyc everything is done, just need to be sure that CI is green. I will leave PR comments a bit later and ping you for review. Next time I will make a PR to be draft, before requesting review

File: cpp/modmesh/view/R3DWidget.cpp
Comment (Tucchhaa) [inline cpp/modmesh/view/R3DWidget.cpp:111]: refactored resetCamera function

File: cpp/modmesh/view/RCameraController.cpp
Comment (Tucchhaa) [inline cpp/modmesh/view/RCameraController.cpp:93]: Changed camera controllers. Now to move it forward/backward need to use Ctrl+W/arrow up and Ctrl+S/arrow down

File: cpp/modmesh/view/RCameraController.cpp
Comment (Tucchhaa) [inline cpp/modmesh/view/RCameraController.cpp:215]: Since keys Q/E and Page Up/Down are not used, I removed keyboard z axis for translation.

File: cpp/modmesh/view/RCameraController.hpp
Comment (Tucchhaa) [inline cpp/modmesh/view/RCameraController.hpp]: Same as `Qt3DExtras::QAbstractCameraController::InputState` but with default values. Need default values when setting values in menu [movement items](https://github.com/solvcon/modmesh/pull/408/files#diff-ea4fe5a63732fb5e650cfceb91cfe707209509f69fa3a50533522e0892aa3014R205).

File: cpp/modmesh/view/RManager.cpp
Comment (Tucchhaa) [inline cpp/modmesh/view/RManager.cpp:277]: View submenu items for camera movement added here

**Comment by Tucchhaa** [comment]
@yyc @tigercosmos ready for review :)

**Comment by yyc** [comment]
I think you're looking for @yungyuc :)

**Comment by tigercosmos** [comment]
> I think you're looking for @yungyuc :)

It's funny 🤣🤣🤣

**Comment by yungyuc** [comment]
> I think you're looking for @yungyuc :)

I can't agree with you more, @yyc

File: cpp/modmesh/view/R3DWidget.cpp
Comment (yungyuc) [reply cpp/modmesh/view/R3DWidget.cpp:111]: I think we should make the reset/original position to be configurable, but I agree setting them from API argument does not look like the best way to do it.  Do you think this is worth of an enhancement ticket, @Tucchhaa ?

File: cpp/modmesh/view/RCameraController.cpp
Comment (yungyuc) [reply cpp/modmesh/view/RCameraController.cpp:93]: I built it and the zoom in/out works with cmd(⌘)+up/down, not ctrl+up/down.  Perhaps the `Qt::Key_Control` means differently on mac?  If so, please document it in code comments (and indicate it in the review comment).

File: cpp/modmesh/view/RManager.cpp
Comment (yungyuc) [reply cpp/modmesh/view/RManager.cpp:277]: I wonder if we should keep all the items in a "folder" (not sure what is the proper name of it) in the menu?

File: cpp/modmesh/view/R3DWidget.cpp
Comment (Tucchhaa) [reply cpp/modmesh/view/R3DWidget.cpp:111]: yes, we can do that. I think we need to expose more camera's props to python: setting camera's position, view center, etc. And we should move all camera's logic to CameraController class. Right now camera's logic is placed on several classes: e.g. camera reset is located in R3DWidget

Comment (Tucchhaa) [reply cpp/modmesh/view/R3DWidget.cpp:111]: That should be a simple enhancement, I can do that in the next PR

File: cpp/modmesh/view/RCameraController.cpp
Comment (Tucchhaa) [reply cpp/modmesh/view/RCameraController.cpp:93]: `Qt::Key_Control` on Windows means Ctrl key, on Mac means Cmd.

File: cpp/modmesh/view/RManager.cpp
Comment (Tucchhaa) [reply cpp/modmesh/view/RManager.cpp:277]: How about "Camera movement" ?

**Comment by Tucchhaa** [comment]
> Would it be difficult to show the keyboard shortcuts of movement in the same way "escape" and "🌐F" is shown?

Yes, that will be difficult. And it will be related to the feature of a menu with shortcuts configuration for camera's movement. But that will require a lot of changes. So I suggest to complete these two things together

File: cpp/modmesh/view/RCameraController.cpp
Comment (tigercosmos) [reply cpp/modmesh/view/RCameraController.cpp:93]: Just curious, is it possible to use the pad with two fingers sliding to zoom in/out? It's also common to use Ctrl+scroll wheel to zoom in/out as well.

Comment (Tucchhaa) [reply cpp/modmesh/view/RCameraController.cpp:93]: That's a good question. Actually I occasionally removed zoom in/out by scroll, so I will bring it back. So, yes, there's a zoom in/out, but without the ctrl and it works with the pad as well

File: cpp/modmesh/view/R3DWidget.cpp
Comment (yungyuc) [reply cpp/modmesh/view/R3DWidget.cpp:111]: Thanks.  I added a bullet under issue #70 to track it.

File: cpp/modmesh/view/RCameraController.cpp
Comment (yungyuc) [reply cpp/modmesh/view/RCameraController.cpp:93]: > `Qt::Key_Control` on Windows means Ctrl key, on Mac means Cmd.

Thanks.  Please add code comments to describe the differences.  It is impossible to tell from pure source code.

File: cpp/modmesh/view/RManager.cpp
Comment (yungyuc) [reply cpp/modmesh/view/RManager.cpp:277]: > How about "Camera movement" ?

Sounds OK, but I'd like to make it shorter.  Could you please call it "Camera move"?

**Comment by yungyuc** [comment]
> > Would it be difficult to show the keyboard shortcuts of movement in the same way "escape" and "🌐F" is shown?
> 
> Yes, that will be difficult. And it will be related to the feature of a menu with shortcuts configuration for camera's movement. But that will require a lot of changes. So I suggest to complete these two things together

It makes sense.  Could you please create a discussion conversion for discussing how we should plan for it?

----

I updated the todo list in [the previous comment](https://github.com/solvcon/modmesh/pull/408#pullrequestreview-2240447427).

**Comment by Tucchhaa** [comment]
@yungyuc @tigercosmos I applied the review. You can review it again

**Review by yungyuc — CHANGES_REQUESTED** [review]
Please follow the [coding style](https://github.com/solvcon/modmesh/blob/master/STYLE.rst).  Many of the ending marks are not added.  There could be other violations.  It could be more productive for you to review yourself.  Please add annotation at your fixes to help reviewers to read.

**Comment by yungyuc** [comment]
@Tucchhaa please add the conversation as discussed in https://github.com/solvcon/modmesh/pull/408#issuecomment-2292850685 .  If it takes time, please make an estimate for when you can create the conversation.

**Comment by Tucchhaa** [comment]
> Please follow the [coding style](https://github.com/solvcon/modmesh/blob/master/STYLE.rst).

I checked my code and I couldn't find any coding styles violations from the doc. 

> Many of the ending marks are not added.

New structs seem to be small enough to omit ending marks. Do you mean to add them to the functions? If so, I suggest then update the [STYLE.rst](https://github.com/solvcon/modmesh/blob/master/STYLE.rst#c-ending-mark) to mention functions too 🙂 

> It could be more productive for you to review yourself.

I always review and double check that everything works, before asking for a review. But I didn't notice that STYLE.rst file, thanks for pointing to that. Next time I will follow that too) 

But imho it would be great that our linter could check all those rules in there. I don't know how much customization clang-format has, but for e.g. eslint (for JS) is highly customizable and almost all these rules can be implemented on linters level. Can we do that?

And I'm a bit confused with functions naming in C++. We have a snake_case and a camelCase. Why is it not only camelCase?

> Please add annotation at your fixes to help reviewers to read

Ok, np

**Comment by Tucchhaa** [comment]
> @Tucchhaa please add the conversation as discussed in [#408 (comment)](https://github.com/solvcon/modmesh/pull/408#issuecomment-2292850685) . If it takes time, please make an estimate for when you can create the conversation.

Yeah, I will add that. Tomorrow evening it will be there

**Comment by Tucchhaa** [comment]
> > @Tucchhaa please add the conversation as discussed in [#408 (comment)](https://github.com/solvcon/modmesh/pull/408#issuecomment-2292850685) . If it takes time, please make an estimate for when you can create the conversation.
> 
> Yeah, I will add that. Tomorrow evening it will be there

@yungyuc the discussion is created: https://github.com/solvcon/modmesh/discussions/412

**Comment by yungyuc** [comment]
> > > @Tucchhaa please add the conversation as discussed in [#408 (comment)](https://github.com/solvcon/modmesh/pull/408#issuecomment-2292850685) . If it takes time, please make an estimate for when you can create the conversation.
> > 
> > 
> > Yeah, I will add that. Tomorrow evening it will be there
> 
> @yungyuc the discussion is created: #412

Sweet.  Thanks!

> > Please follow the [coding style](https://github.com/solvcon/modmesh/blob/master/STYLE.rst).
> 
> I checked my code and I couldn't find any coding styles violations from the doc.
> 
> > Many of the ending marks are not added.
> 
> New structs seem to be small enough to omit ending marks. Do you mean to add them to the functions? If so, I suggest then update the [STYLE.rst](https://github.com/solvcon/modmesh/blob/master/STYLE.rst#c-ending-mark) to mention functions too 🙂

Small classes in the beginning does not guarantee the compactness later.  Usually, they will end up with long classes.  When functions grow longer, they should be refactored to smaller pieces.  But it is usually not the case for classes.  It is not common to be able to break a class into smaller ones.

> > It could be more productive for you to review yourself.
> 
> I always review and double check that everything works, before asking for a review. But I didn't notice that STYLE.rst file, thanks for pointing to that. Next time I will follow that too)

I wasn't clear enough myself.  I fully understand that you have reviewed before requesting for review.  The two sentences should be interpreted together: "It could be more productive for you to review yourself. Please add annotation at your fixes to help reviewers to read."

What I wanted to emphasize is to add annotations when you review the code yourself.  It helps the reviewers a lot.  It also help future maintainers when they need to do git archeology 10 years after.

> But imho it would be great that our linter could check all those rules in there. I don't know how much customization clang-format has, but for e.g. eslint (for JS) is highly customizable and almost all these rules can be implemented on linters level. Can we do that?

It is.  I tried before but couldn't make it.  Feel free to try again.  But it may be hard.  If you are not super enthusiastic on it, don't spend too much time.

For the detail style requirements, simply getting familiar with the coding style and pay attention with the code base are more productive, in my experience.

> And I'm a bit confused with functions naming in C++. We have a snake_case and a camelCase. Why is it not only camelCase?

I want `snake_case()` because of the preference to use [single-name accessors](https://github.com/solvcon/modmesh/blob/master/STYLE.rst#c-encapsulation).  It more aligns with the Python attribute accessor style (the [descriptor protocol](https://docs.python.org/3/howto/descriptor.html)).

> > Please add annotation at your fixes to help reviewers to read
> 
> Ok, np

File: cpp/modmesh/view/RManager.cpp
Comment (Tucchhaa) [inline cpp/modmesh/view/RManager.cpp:373]: For orbit camera axisValue is not 1.0f (as in FPS camera), because rotation is very small.

File: cpp/modmesh/view/RCameraController.cpp
Comment (Tucchhaa) [inline cpp/modmesh/view/RCameraController.cpp:303]: Just a refactoring.
removed check for `input.leftMouseButtonActive` because it's always false here. Because in the [if above](https://github.com/solvcon/modmesh/pull/408/commits/ee98cd1c5d4680f9c76ec5c98a45f71a989e7a88#diff-be9511fa234431d29c196ed57eaf68dbf67a702bc5400c605cdb9d5450a48d33L263), there's already a check for `input.leftMouseButtonActive`

File: cpp/modmesh/view/RManager.cpp
Comment (yungyuc) [inline cpp/modmesh/view/RManager.cpp:285]: The order of the movement is wsda.  It is fine, and I like it to align with the Cartesian coordinate direction (w: +y, s:-y, d: +x, a: -x).

But the convention in literature of the directional keys are wsad or wasd.  This is a little bit unintuitive.  Googling "wsda" does not return anything related to the direction keys.

Having said that, I do not want to change the order of wsda.  This note is only for our information.

File: cpp/modmesh/view/RManager.cpp
Comment (yungyuc) [reply cpp/modmesh/view/RManager.cpp:373]: We will need to experience this later after merging.  It's great that you are starting to fine-tune.

File: cpp/modmesh/view/RCameraController.cpp
Comment (yungyuc) [reply cpp/modmesh/view/RCameraController.cpp:303]: Code like this needs a lot of fine-tuning.  Thanks for doing this.  It's great.

**Review by yungyuc — APPROVED** [review]
LGTM.

Please remember to leave a global comment when requesting for review.  Following the convention will help future maintainers.  It is not clear to just click the request review button/link.

**Pull request merged**
